File: app/src/main/java/fr/dmconcept/bob/BobApplication.java
package fr.dmconcept.bob;

import android.app.Application;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;
import fr.dmconcept.bob.models.BoardConfig;
import fr.dmconcept.bob.models.Project;
import fr.dmconcept.bob.models.ServoConfig;
import fr.dmconcept.bob.models.Step;
import fr.dmconcept.bob.models.dao.BoardConfigDao;
import fr.dmconcept.bob.models.dao.ProjectDao;
import fr.dmconcept.bob.models.helpers.BobSqliteOpenHelper;

import java.util.ArrayList;

public class BobApplication extends Application {

    private static final String TAG = "BobApplication";

    BobSqliteOpenHelper mOpenHelper;
    SQLiteDatabase      mDatabase;

    BoardConfigDao mBoardConfigDao;
    ProjectDao mProjectsDao;

    // The server IP address
    String mServerIP;

    @Override
    public void onCreate() {

        Log.i(TAG, "onCreate()");

        mOpenHelper = new BobSqliteOpenHelper(this);
        mDatabase   = mOpenHelper.getWritableDatabase();

        mBoardConfigDao = new BoardConfigDao(mDatabase);
        mProjectsDao    = new ProjectDao(mDatabase, mBoardConfigDao);

        if (BuildConfig.DEBUG && mProjectsDao.findAll().size() == 0)
            // First run, create the fixtures
            createFixtures();

        super.onCreate();

    }

    private void createFixtures() {

        Log.i(TAG, "DEBUG MODE - Creating servos config fixtures");

        ArrayList<ServoConfig> servoConfigs1 = new ArrayList<ServoConfig>() {{
            add(new ServoConfig(3, 558, 2472, 50));
            add(new ServoConfig(4, 558, 2472, 50));
        }};

        ArrayList<ServoConfig> servoConfigs2 = new ArrayList<ServoConfig>() {{
            add(new ServoConfig(3, 558, 2472, 50));
            add(new ServoConfig(4, 558, 2472, 50));
            add(new ServoConfig(5, 558, 2472, 50));
            add(new ServoConfig(6, 558, 2472, 50));
            add(new ServoConfig(7, 558, 2472, 50));
            add(new ServoConfig(8, 558, 2472, 50));
            add(new ServoConfig(9, 558, 2472, 50));
        }};

        BoardConfig config1 = new BoardConfig("Servos on pins 3 and 4"  , servoConfigs1);
        BoardConfig config2 = new BoardConfig("Servos on pins 3 to 9", servoConfigs2);

        config1.setId(mBoardConfigDao.save(config1));
        config2.setId(mBoardConfigDao.save(config2));

        Log.i(TAG, "DEBUG MODE - Creating project...");

        // Project 1
        mProjectsDao.create(
            new Project(-1, "Simple demo project", config1, new ArrayList<Step>() {{
                add(new Step(4000, new ArrayList<Integer>() {{ add( 1); add(50); }}));
                add(new Step(2000, new ArrayList<Integer>() {{ add(99); add(50); }}));
                add(new Step(4000, new ArrayList<Integer>() {{ add(50); add(50); }}));
                add(new Step(0   , new ArrayList<Integer>() {{ add( 1); add(50); }}));
            }})
        );

        // Project 2
        mProjectsDao.create(
            new Project(-1, "Bob demo project", config2, new ArrayList<Step>() {{
                add(new Step(5000, new ArrayList<Integer>() {{
                    add(  0); add(20); add(20); add(10); add(20); add(80); add( 100);
                }}));
                add(new Step(6000, new ArrayList<Integer>() {{
                    add( 25); add(40); add(40); add(20); add(40); add(60); add( 50);
                }}));
                add(new Step(5000, new ArrayList<Integer>() {{
                    add( 75); add(60); add(60); add(30); add(60); add(40); add( 25);
                }}));
                add(new Step(10000, new ArrayList<Integer>() {{
                    add(100); add(80); add(80); add(40); add(80); add(20); add(  0);
                }}));
            }})
        );

        Log.i(TAG, "DEBUG MODE - Fixtures creation done");

    }

    public ProjectDao getProjectsDao() {
        return mProjectsDao;
    }

    public BoardConfigDao getBoardConfigDao() {
        return mBoardConfigDao;
    }

    public String getServerIP() {
        return mServerIP;
    }

    public void setServerIP(String ip) {
        Log.i(TAG, "setServerIP(" + ip + ")");
        mServerIP = ip;
    }

    @Override
    public void onTerminate() {

        // Close the database and the DB helper
        mDatabase.close();
        mOpenHelper.close();

        super.onTerminate();

    }


}

File: bob-client/src/com/protogenefactory/ioiomaster/client/BobApplication.scala
package com.protogenefactory.ioiomaster.client

import java.util.UUID

import android.app.Application
import android.database.sqlite.SQLiteDatabase
import com.protogenefactory.ioiomaster.client.BobApplication.Preferences
import com.protogenefactory.ioiomaster.client.connections.{Connection, LocalConnection, RemoteConnection}
import com.protogenefactory.ioiomaster.client.models.dao.{BoardConfigDao, ProjectDao}
import com.protogenefactory.ioiomaster.client.models.helpers.BobSqliteOpenHelper
import com.protogenefactory.ioiomaster.client.models.{BoardConfig, Project, ServoConfig, Step}
import com.protogenefactory.ioiomaster.server.services.ServerService
import org.scaloid.common._

object BobApplication {

  object Preferences {
    val SERVER_IP = "serverIP"
    val AUTOPLAY  = "autoplay"
  }

  object Params {
    val PORT = 8000
  }

}

class BobApplication extends Application with SContext {

  implicit override val loggerTag = LoggerTag("Bob")

  lazy val sqliteOpenHelper : BobSqliteOpenHelper = new BobSqliteOpenHelper(this)
  lazy val sqliteDatabase   : SQLiteDatabase      = sqliteOpenHelper.getWritableDatabase

  lazy val boardConfigDao   : BoardConfigDao      = new BoardConfigDao(sqliteDatabase)
  lazy val projectsDao      : ProjectDao          = new ProjectDao    (sqliteDatabase, boardConfigDao)

  var connection: Connection = null

  override def onCreate() {

    info("BobApplication.onCreate()")

    if (/*BuildConfig.DEBUG && */ projectsDao.isEmpty)
      firstRun()

    // Start the server service
    startService[ServerService]

    super.onCreate()

  }

  private def firstRun() {

    info("BobApplication.onCreate() First run, creating the servos config fixtures")

    def servoConfigs(servoPins: List[Int]): Vector[ServoConfig] = servoPins.map { s =>
      ServoConfig(s"Servo $s", s)
    }.toVector

    val configs = List(
      BoardConfig("Servos on 2 first pins", servoConfigs(List(3, 4)               )),
      BoardConfig("Servos on all pins"    , servoConfigs(List(3, 4, 5, 6, 7, 10, 11, 12, 13)))
    )

    configs.foreach(boardConfigDao.create)

    info("BobApplication.onCreate() First run, creating the project fixtures")

    // Project 1
    projectsDao.create(
      Project(UUID.randomUUID.toString, "Simple demo project", configs(0), Vector(
        Step(4000, Vector( 1, 100)),
        Step(2000, Vector(99, 1)),
        Step(4000, Vector(50, 100)),
        Step(0   , Vector( 1, 1))
      ))
    )

    // Project 2
    projectsDao.create(
      Project(UUID.randomUUID.toString, "Bob demo project", configs(1), Vector(
        Step( 500, Vector(  0, 100,   0, 100,   0,  40,   0,  0,  40)),
        Step(1000, Vector(100,   0,  20,  80, 100,  60,  20, 20,  60)),
        Step( 200, Vector(  0, 100,  40,  60,  20,  20,  40, 40,  20)),
        Step( 500, Vector(100,   0,  60,  40,  80,  80,  60, 60,  80)),
        Step( 200, Vector(  0, 100,  80,  20,  40,   0,  80, 80,   0)),
        Step( 500, Vector(  0, 100,   0, 100,   0,  40,   0,  0,  40)),
        Step(1000, Vector(100,   0,  20,  80, 100,  60,  20, 20,  60)),
        Step( 200, Vector(  0, 100,  40,  60,  20,  20,  40, 40,  20)),
        Step( 500, Vector(100,   0,  60,  40,  80,  80,  60, 60,  80)),
        Step( 200, Vector(  0, 100,  80,  20,  40,   0,  80, 80,   0)),
        Step(  0 , Vector(100,   0, 100,   0,  60, 100, 100,  0, 100))
      ))
    )

  }

  def setLocalConnection(serverService: LocalServiceConnection[ServerService]) {
    info("BobApplication.setLocalConnection()")
    connection = new LocalConnection(serverService)
  }

  def setRemoteConnection(remoteIP: String) {

    info(s"BobApplication.setRemoteConnection($remoteIP)")

    defaultSharedPreferences
      .edit
      .putString(Preferences.SERVER_IP, remoteIP)
      .apply()

    connection = new RemoteConnection(remoteIP)

  }

  def serverIP = defaultSharedPreferences.getString(Preferences.SERVER_IP, "")

}
